wled00/button.cpp

```diff
@@ -186,61 +227,43 @@ void handleButton()
     }
 
     //momentary button logic
-    if (isButtonPressed(b)) //pressed
-    {
+    if (isButtonPressed(b)) { //pressed
+
       if (!buttonPressedBefore[b]) buttonPressedTime[b] = millis();
       buttonPressedBefore[b] = true;
 
-      if (millis() - buttonPressedTime[b] > 600) //long press
-      {
-        if (!buttonLongPressed[b]) 
-        {
-          if (macroLongPress[b]) {applyPreset(macroLongPress[b], CALL_MODE_BUTTON);}
-          else _setRandomColor(false,true);
-
-          // publish MQTT message
-          if (buttonPublishMqtt && WLED_MQTT_CONNECTED) {
-            char subuf[64];
-            sprintf_P(subuf, _mqtt_topic_button, mqttDeviceTopic, (int)b);
-            mqtt->publish(subuf, 0, false, "long");
-          }
-
-          buttonLongPressed[b] = true;
+      if (millis() - buttonPressedTime[b] > 600) { //long press
+        if (!buttonLongPressed[b]) longPressAction(b);
+        else if (b) { // repeatable action (~3 times per s) on button > 0
+          longPressAction(b);
+          buttonPressedTime[b] = millis() - 300; // 300ms
         }
+        buttonLongPressed[b] = true;
       }
-    }
-    else if (!isButtonPressed(b) && buttonPressedBefore[b]) //released
-    {
+
+    } else if (!isButtonPressed(b) && buttonPressedBefore[b]) { //released
+
       long dur = millis() - buttonPressedTime[b];
       if (dur < WLED_DEBOUNCE_THRESHOLD) {buttonPressedBefore[b] = false; continue;} //too short "press", debounce
-      bool doublePress = buttonWaitTime[b];
+      bool doublePress = buttonWaitTime[b]; //did we have short press before?
       buttonWaitTime[b] = 0;
 
-      if (dur > 6000 && b==0) //long press on button 0
-      {
+      if (b == 0 && dur > 6000) { //long press on button 0 (when released)
         WLED::instance().initAP(true);
-      }
-      else if (!buttonLongPressed[b]) { //short press
-        if (macroDoublePress[b])
-        {
-          if (doublePress) {
-            applyPreset(macroDoublePress[b], CALL_MODE_BUTTON);
-  
-            // publish MQTT message
-            if (buttonPublishMqtt && WLED_MQTT_CONNECTED) {
-              char subuf[64];
-              sprintf_P(subuf, _mqtt_topic_button, mqttDeviceTopic, (int)b);
-              mqtt->publish(subuf, 0, false, "double");
-            }
-          } else buttonWaitTime[b] = millis();
-        } else shortPressAction(b);
```

Comment:
I appreciate the extra usermod hook and the `button.cpp()` cleanup by moving the actions into separate functions :)

Just one minor nitpick:
Rather than a default double press button action (set as toggle on button 0, cycle palettes on button IDs >0), for UX in my opinion it would be preferable to fire the short press action immediately upon release if no double press macro is defined. Otherwise, users who don't wish to use double press will still have a 350ms delay on every press until the short press action is fired.

I have changed this back just for button 0 right now (since its default double press action is just on/off, same as the short press action anyways), on buttons >0 it is debatable if the default double press action (palette cycle) or the potential for instant short press action is more desirable, so I'd be fine with leaving it as-is for now.

Reply:
Agreed on all topics.